wtransport/src/config.rs

```diff
@@ -728,7 +728,55 @@ impl ClientConfigBuilder<states::WantsRootStore> {
         })
     }
 
-    fn native_cert_store() -> RootCertStore {
+    /// Configures the client to skip *some* server certificates validation.
+    ///
+    /// This method configures the client to accept server certificates
+    /// whose digests match the specified *SHA-256* hashes and fulfill
+    /// some additional constraints (*see notes below*).
+    ///
+    /// This is useful for scenarios where clients need to accept known
+    /// self-signed certificates or certificates from non-standard authorities.
+    ///
+    /// This method configuration is similar to the
+    /// [browser W3C WebTransport API](https://www.w3.org/TR/webtransport/#dom-webtransportoptions-servercertificatehashes).
+    ///
+    /// # Notes
+    ///
+    /// - The current time MUST be within the validity period of the certificate.
+    /// - The total length of the validity period MUST NOT exceed *two* weeks.
+    /// - Only certificates for which the public key algorithm is *ECDSA* with the *secp256r1* are accepted.
+    #[cfg(all(feature = "dangerous-configuration", feature = "self-signed"))]
+    #[cfg_attr(
+        docsrs,
+        doc(cfg(all(feature = "dangerous-configuration", feature = "self-signed")))
+    )]
+    pub fn with_server_certificate_hashes<I>(
+        self,
+        hashes: I,
+    ) -> ClientConfigBuilder<states::WantsTransportConfigClient>
+    where
+        I: IntoIterator<Item = crate::tls::Sha256Digest>,
+    {
+        let mut tls_config = Self::build_tls_config(Arc::new(RootCertStore::empty()));
+
+        let root_store = Self::native_cert_store();
```

Comment:
Is this correct? I'd assume native certs must be ignored, no?

Reply:
It depends. This is not webtransport standard, rather an inspiration from W3C API.

In particular here: https://www.w3.org/TR/webtransport/#dom-webtransportoptions-servercertificatehashes

> the user agent SHALL deem a server certificate trusted if and only if it can successfully verify a certificate hash against serverCertificateHashes and satisfies custom certificate requirements.

Where "custom certificate requirements" opens possibilities for custom implementations:
> The user agent MAY impose additional implementation-defined requirements on the certificate.

---

However, the expression "additional requirements" makes me think more to a logical "AND" rather than a "OR".

I assume that certificates validated by proper PKI mechanism are "safer" than this hashes; however, I guess we might remote those in order to give the client of this API full control and accept strictly what is specified in the hashes.

Thanks for the feedback!

Reply:
https://github.com/BiagioFesta/wtransport/pull/133